### PR TITLE
Add C++ sample robot code

### PR DESCRIPTION
**Pull request by prateekma**
Pretty much translated the Java equivalent to C++ to the best of my abilities. Haven't tested this on a robot yet, but will do so when I get access to one of our robots.

Let me know if there are any errors that I should fix.

### CODE REVIEW

**Comment by virtuald** [comment]
Since we distribute this to high school students, I'd prefer that someone at least verifies that it compiles before we merge it.

File: robot-java/src/main/java/dc/Robot.java [hunk 1]
```diff
@@ -157,7 +154,7 @@ public void autonomousPeriodic() {
 		priorAutospeed = autospeed;
 
 		// command motors to do things
-drive.tankDrive(autospeed, autospeed, false);
+		drive.tankDrive(autospeed, autospeed, false);
```
Comment (virtuald) [inline robot-java/src/main/java/dc/Robot.java]: Why did this file change? It's not relevant to your PR.

**Comment by auscompgeek** [comment]
> Since we distribute this to high school students, I'd prefer that someone at least verifies that it compiles before we merge it.

It seems like we should have these sorts of checks in CI.

File: robot-java/src/main/java/dc/Robot.java (hunk 1 above)
Comment (prateekma) [reply robot-java/src/main/java/dc/Robot.java]: My bad, I probably auto-formatted it by accident; I'll revert this change.

**Comment by prateekma** [comment]
I've verified that both projects compile by running a `./gradlew build`. I will test out the actual code on a robot tonight if I get the chance.

**Comment by prateekma** [comment]
I ran the `robot-cpp-talonsrx` on our robot tonight and found one bug that I fixed and committed. It involved the follower TalonSRX's not being added to SpeedControllerGroup. I thought this wasn't needed because they would follow the leader, but apparently not.

After that change was made, the program worked as intended with `data_logger.py`. I was unable to test out the version with Sparks since we don't have a robot that has Sparks on it. Fortunately, I added the follower motors to the speed controller group in that version of the program.

**Comment by auscompgeek** [comment]
> It involved the follower TalonSRX's not being added to SpeedControllerGroup. I thought this wasn't needed because they would follow the leader, but apparently not.

I think you've debugged this the wrong way. In the same commit, you kept the object around, whereas originally you had an object that got destructed after RobotInit. Following a Talon SRX in PercentOutput mode should work without any problems.

**Comment by auscompgeek** [comment]
Looking at the Java and Python versions, it's quite likely they also suffer from this bug, assuming the Java and Python bindings actually call the CCI destructor as well. (I suspect the Python one doesn't currently, which itself is a bug.)

**Comment by prateekma** [comment]
Hmm, looking back in hindsight, you're right. The follower objects were destructed after RobotInit. This probably caused the bug I described.

I was a little confused because we've done similar things in Kotlin (JVM so basically the same as Java) i.e. declaring follower objects and setting them to follow the master motor in a configuration method. In this case, even when the method ends, the follower motor still continues to follow the motor, even though the master is the only object being kept around.

**Comment by auscompgeek** [comment]
It's very much possible that the Phoenix Java library doesn't call the underlying destructor. I haven't taken a look at the Java bits this year yet (mostly on account of this year's stuff not being on GitHub until recently).

**Comment by Oblarg** [comment]
Could you update this to reflect the new repo structure and include versions for arm characterization?

**Comment by prateekma** [comment]
I will test the `robot-cpp-talonsrx` arm characterization on our robot soon (hopefully next week) to make sure that it works; but I simply adapted the `robot-cpp-talonsrx` drive characterization to reflect the changes needed to make it work with arm characterization.

File: .gitignore [hunk 2]
```diff
@@ -7,7 +7,7 @@ bin
...
 
-*/.gradle
-*/.vscode
-*/.wpilib
-*/build
\ No newline at end of file
+**/.gradle
+**/.vscode
+**/.wpilib
+**/build
```
Comment (auscompgeek) [inline .gitignore]: I think if we leave off the `**/` it'll work as expected too.

File: README.md [hunk 3]
```diff
@@ -13,8 +13,12 @@ Included Characterization Tools
...
 The robot characterization toolsuite currently supports chracterization for:
 
-- Drivetrains
-- Arms
+* [Java](robot-java)
+* [Java TalonSRX](robot-java-talonsrx)
+* [Python](robot-python)
+* [Python TalonSRX](robot-python-talonsrx)
+* [C++](robot-cpp)
+* [C++ TalonSRX](robot-cpp-talonsrx)
```
Comment (auscompgeek) [inline README.md]: I think you want to change this only under the subdirectory readmes.

**Comment by auscompgeek** [comment]
Would it be better to have the number array as a local variable in AutonomousPeriodic? It's only used there and would avoid the length getting out of sync if anyone decides to send more data.

File: .gitignore (hunk 2 above)
Comment (prateekma) [reply .gitignore]: Hmm, without the second asterisk, Git resolves those folders as untracked. With the double asterisks, it ignores them.

This screenshot was with a single asterisk: http://prntscr.com/oaloid

File: README.md (hunk 3 above)
Comment (prateekma) [reply README.md]: Whoops, my bad. Fixed this in latest commit.

**Comment by prateekma** [comment]
I can make the number array a static variable in AutonomousPeriodic. This will make it local, but it won't need to be constructed and destroyed every 20 milliseconds.

**Comment by auscompgeek** [comment]
> This will make it local, but it won't need to be constructed and destroyed every 20 milliseconds.

A trivial allocation on the stack has no runtime cost.

If we have an actual local variable, the array can be initialised on declaration, and I believe we can omit the size in that case.

File: .gitignore (hunk 2 above)
Comment (auscompgeek) [reply .gitignore]: That's why I'm saying leave off the globs entirely.

Comment (prateekma) [reply .gitignore]: Ah yes, that does work. Fixed in 4752beb

**Comment by prateekma** [comment]
Tested the C++ code on our robot today and everything works as expected

**Comment by virtuald** [comment]
Thank you for your contribution! Sorry for the delay.

**Pull request merged**